sql: running SQL PREPARE and EXECUTE via the extended protocol panics the server

If you run a SQL-level PREPARE statement (e.g. `PREPARE x AS SELECT `) via the extended protocol, the server panics with the following stack trace:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x70 pc=0x2fb7ad4]

goroutine 903 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00150c600, {0x523ecf0, 0xc0028bf6c0}, {0x3b14260, 0x756e3d0})
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1043 +0x1a5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:744 +0x53
panic({0x3b14260, 0x756e3d0})
        GOROOT/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/sql/types.(*T).Oid(...)
        github.com/cockroachdb/cockroach/pkg/sql/types/pkg/sql/types/types.go:1169
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).addPreparedStmt(0xc00150c600, {0x523ed98, 0xc002c476b0}, {0xc000fd9eba, 0x0}, {{{0x526c278, 0xc002082380}, {0xc000fd9ebd, 0x4d}, 0x2, ...}, ...}, ...)
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:127 +0x514
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPrepare(0xc00150c600, {0x523ed98, 0xc002c47680}, {{0xc000fd9eba, 0x2}, {{0x526c278, 0xc002082380}, {0xc000fd9ebd, 0x4d}, 0x2, ...}, ...})
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_prepare.go:63 +0x3b8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc00150c600)
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1948 +0x10c6
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc00150c600, {0x523ecf0, 0xc0028bf6c0}, 0xc00223d920, {0xc00045b110, 0xc00343bdf8, 0xc002047550, 0x0}, 0xc00070bb30)
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1779 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc001c66000, {0x523ecf0, 0xc0028bf6c0}, {0xc0003dd100}, {0x3, 0xc000a39190, 0xc002047550, 0x0}, 0xc0028bf640)
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:746 +0xa8
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
        github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:723 +0x3b0
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
        github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:642 +0x272
```

There's a similar crash that's possible with a SQL-level `EXECUTE` on a SQL-level prepared statement with type hints if you try to run that via the extended protocol:

```
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +runtime error: index out of range [1] with length 1
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +(1) attached stack trace
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  -- stack trace:
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:744
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | [...repeated from below...]
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +Wraps: (2) while executing: EXECUTE _ ('_', '_')
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +Wraps: (3) attached stack trace
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  -- stack trace:
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:744
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | runtime.gopanic
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  GOROOT/src/runtime/panic.go:1038
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | runtime.goPanicIndex
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  GOROOT/src/runtime/panic.go:90
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).bufferBatch
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:1379
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).AddBatch.func1
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:235
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).addInternal
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:212
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).AddBatch
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:233
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).PushBatch
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1155
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:299
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:259
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:587
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1423
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1431
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1108
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:683
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:141
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:229
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1924
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1926
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1779
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:746
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  |  github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:723
E220225 22:28:09.574864 906 1@util/log/logcrash/crash_reporting.go:174  [n1,client=127.0.0.1:40780,user=root] 325 +  | runtime.goexit
```

--- a/pkg/sql/BUILD.bazel
+++ b/pkg/sql/BUILD.bazel
@@ -536,6 +536,7 @@ go_test(
         "sort_test.go",
         "split_test.go",
         "sql_cursor_test.go",
+        "sql_prepare_test.go",
         "statement_mark_redaction_test.go",
         "table_ref_test.go",
         "table_test.go",

--- a/pkg/sql/BUILD.bazel
+++ b/pkg/sql/BUILD.bazel
@@ -535,6 +535,7 @@ go_test(
         "show_trace_replica_test.go",
         "sort_test.go",
         "split_test.go",
+        "sql_cursor_test.go",
         "statement_mark_redaction_test.go",
         "table_ref_test.go",
         "table_test.go",

--- a/pkg/sql/conn_executor_prepare.go
+++ b/pkg/sql/conn_executor_prepare.go
@@ -272,10 +272,9 @@ func (ex *connExecutor) populatePrepared(
 	// only allows SELECT, INSERT, UPDATE, DELETE and VALUES statements to be
 	// prepared.
 	// See: https://www.postgresql.org/docs/current/static/sql-prepare.html
-	// However, we allow a large number of additional statements.
-	// As of right now, the optimizer only works on SELECT statements and will
-	// fallback for all others, so this should be safe for the foreseeable
-	// future.
+	// However, we must be able to handle every type of statement below because
+	// the Postgres extended protocol requires running statements via the prepare
+	// and execute paths.
 	flags, err := p.prepareUsingOptimizer(ctx)
 	if err != nil {
 		log.VEventf(ctx, 1, "optimizer prepare failed: %v", err)
@@ -553,7 +552,20 @@ func (ex *connExecutor) execDescribe(
 
 		res.SetInferredTypes(ps.InferredTypes)
 
-		if stmtHasNoData(ps.AST) {
+		ast := ps.AST
+		if execute, ok := ast.(*tree.Execute); ok {
+			// If we're describing an EXECUTE, we need to look up the statement type
+			// of the prepared statement that the EXECUTE refers to, or else we'll
+			// return the wrong information for describe.
+			innerPs, found := ex.extraTxnState.prepStmtsNamespace.prepStmts[string(execute.Name)]
+			if !found {
+				return retErr(pgerror.Newf(
+					pgcode.InvalidSQLStatementName,
+					"unknown prepared statement %q", descCmd.Name))
+			}
+			ast = innerPs.AST
+		}
+		if stmtHasNoData(ast) {
 			res.SetNoDataRowDescription()
 		} else {
 			res.SetPrepStmtOutput(ctx, ps.Columns)

--- a/pkg/sql/conn_executor_prepare.go
+++ b/pkg/sql/conn_executor_prepare.go
@@ -160,6 +160,17 @@ func (ex *connExecutor) prepare(
 		return prepared, nil
 	}
 
+	origNumPlaceholders := stmt.NumPlaceholders
+	switch stmt.AST.(type) {
+	case *tree.Prepare:
+		// Special case: we're preparing a SQL-level PREPARE using the
+		// wire protocol. There's an ambiguity from the perspective of this code:
+		// any placeholders that are inside of the statement that we're preparing
+		// shouldn't be treated as placeholders to the PREPARE statement. So, we
+		// edit the NumPlaceholders field to be 0 here.
+		stmt.NumPlaceholders = 0
+	}
+
 	var flags planFlags
 	prepare := func(ctx context.Context, txn *kv.Txn) (err error) {
 		p := &ex.planner
@@ -205,6 +216,9 @@ func (ex *connExecutor) prepare(
 			},
 		}
 		prepared.Statement = stmt.Statement
+		// When we set our prepared statement, we need to make sure to propagate
+		// the original NumPlaceholders if we're preparing a PREPARE.
+		prepared.Statement.NumPlaceholders = origNumPlaceholders
 		prepared.StatementNoConstants = stmt.StmtNoConstants
 		prepared.StatementSummary = stmt.StmtSummary
 

--- a/pkg/sql/plan_opt.go
+++ b/pkg/sql/plan_opt.go
@@ -53,7 +53,7 @@ func (p *planner) prepareUsingOptimizer(ctx context.Context) (planFlags, error) 
 	opc := &p.optPlanningCtx
 	opc.reset()
 
-	switch stmt.AST.(type) {
+	switch t := stmt.AST.(type) {
 	case *tree.AlterIndex, *tree.AlterTable, *tree.AlterSequence,
 		*tree.Analyze,
 		*tree.BeginTransaction,
@@ -64,7 +64,6 @@ func (p *planner) prepareUsingOptimizer(ctx context.Context) (planFlags, error) 
 		*tree.CreateStats,
 		*tree.Deallocate, *tree.Discard, *tree.DropDatabase, *tree.DropIndex,
 		*tree.DropTable, *tree.DropView, *tree.DropSequence, *tree.DropType,
-		*tree.Execute,
 		*tree.Grant, *tree.GrantRole,
 		*tree.Prepare,
 		*tree.ReleaseSavepoint, *tree.RenameColumn, *tree.RenameDatabase,
@@ -79,6 +78,23 @@ func (p *planner) prepareUsingOptimizer(ctx context.Context) (planFlags, error) 
 		// optbuilder so they would error out. Others (like CreateIndex) have planning
 		// code that can introduce unnecessary txn retries (because of looking up
 		// descriptors and such).
+		return opc.flags, nil
+
+	case *tree.Execute:
+		// This statement is going to execute a prepared statement. To prepare it,
+		// we need to set the expected output columns to the output columns of the
+		// prepared statement that the user is trying to execute.
+		name := string(t.Name)
+		prepared, ok := p.preparedStatements.Get(name)
+		if !ok {
+			// We're trying to prepare an EXECUTE of a statement that doesn't exist.
+			// Let's just give up at this point.
+			// Postgres doesn't fail here, instead it produces an EXECUTE that returns
+			// no columns. This seems like dubious behavior at best.
+			return opc.flags, pgerror.Newf(pgcode.UndefinedPreparedStatement,
+				"no such prepared statement %s", name)
+		}
+		stmt.Prepared.Columns = prepared.Columns
 		return opc.flags, nil
 
 	case *tree.ExplainAnalyze:

--- a/pkg/sql/sql_cursor.go
+++ b/pkg/sql/sql_cursor.go
@@ -30,10 +30,6 @@ import (
 // DeclareCursor implements the DECLARE statement.
 // See https://www.postgresql.org/docs/current/sql-declare.html for details.
 func (p *planner) DeclareCursor(ctx context.Context, s *tree.DeclareCursor) (planNode, error) {
-	if p.autoCommit {
-		return nil, pgerror.Newf(pgcode.NoActiveSQLTransaction, "DECLARE CURSOR can only be used in transaction blocks")
-	}
-
 	if s.Hold {
 		return nil, unimplemented.NewWithIssue(41412, "DECLARE CURSOR WITH HOLD")
 	}
@@ -44,61 +40,71 @@ func (p *planner) DeclareCursor(ctx context.Context, s *tree.DeclareCursor) (pla
 		return nil, unimplemented.NewWithIssue(41412, "DECLARE SCROLL CURSOR")
 	}
 
-	ie := p.ExecCfg().InternalExecutorFactory(ctx, p.SessionData())
-	cursorName := s.Name.String()
-	if cursor, _ := p.sqlCursors.getCursor(cursorName); cursor != nil {
-		return nil, pgerror.Newf(pgcode.DuplicateCursor, "cursor %q already exists", cursorName)
-	}
+	return &delayedNode{
+		name: s.String(),
+		constructor: func(ctx context.Context, p *planner) (_ planNode, _ error) {
+			if p.extendedEvalCtx.TxnImplicit {
+				return nil, pgerror.Newf(pgcode.NoActiveSQLTransaction, "DECLARE CURSOR can only be used in transaction blocks")
+			}
 
-	// Try to plan the cursor query to make sure that it's valid.
-	stmt := makeStatement(parser.Statement{AST: s.Select}, ClusterWideID{})
-	pt := planTop{}
-	pt.init(&stmt, &p.instrumentation)
-	opc := &p.optPlanningCtx
-	opc.p.stmt = stmt
-	opc.reset()
+			ie := p.ExecCfg().InternalExecutorFactory(ctx, p.SessionData())
+			cursorName := s.Name.String()
+			if cursor, _ := p.sqlCursors.getCursor(cursorName); cursor != nil {
+				return nil, pgerror.Newf(pgcode.DuplicateCursor, "cursor %q already exists", cursorName)
+			}
 
-	memo, err := opc.buildExecMemo(ctx)
-	if err != nil {
-		return nil, err
-	}
-	if err := opc.runExecBuilder(
-		&pt,
-		&stmt,
-		newExecFactory(p),
-		memo,
-		p.EvalContext(),
-		p.autoCommit,
-	); err != nil {
-		return nil, err
-	}
-	if pt.flags.IsSet(planFlagContainsMutation) {
-		// Cursors with mutations are invalid.
-		return nil, pgerror.Newf(pgcode.FeatureNotSupported,
-			"DECLARE CURSOR must not contain data-modifying statements in WITH")
-	}
+			// Try to plan the cursor query to make sure that it's valid.
+			stmt := makeStatement(parser.Statement{AST: s.Select}, ClusterWideID{})
+			pt := planTop{}
+			pt.init(&stmt, &p.instrumentation)
+			opc := &p.optPlanningCtx
+			opc.p.stmt = stmt
+			opc.reset()
 
-	statement := s.Select.String()
-	rows, err := ie.QueryIterator(ctx, "sql-cursor", p.txn, statement)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to DECLARE CURSOR")
-	}
-	inputState := p.txn.GetLeafTxnInputState(ctx)
-	cursor := &sqlCursor{
-		InternalRows: rows,
-		readSeqNum:   inputState.ReadSeqNum,
-		txn:          p.txn,
-		statement:    statement,
-		created:      timeutil.Now(),
-	}
-	if err := p.sqlCursors.addCursor(cursorName, cursor); err != nil {
-		// This case shouldn't happen because cursor names are scoped to a session,
-		// and sessions can't have more than one statement running at once. But
-		// let's be diligent and clean up if it somehow does happen anyway.
-		_ = cursor.Close()
-		return nil, err
-	}
-	return newZeroNode(nil /* columns */), nil
+			memo, err := opc.buildExecMemo(ctx)
+			if err != nil {
+				return nil, err
+			}
+			if err := opc.runExecBuilder(
+				&pt,
+				&stmt,
+				newExecFactory(p),
+				memo,
+				p.EvalContext(),
+				p.autoCommit,
+			); err != nil {
+				return nil, err
+			}
+			if pt.flags.IsSet(planFlagContainsMutation) {
+				// Cursors with mutations are invalid.
+				return nil, pgerror.Newf(pgcode.FeatureNotSupported,
+					"DECLARE CURSOR must not contain data-modifying statements in WITH")
+			}
+
+			statement := s.Select.String()
+			itCtx := context.Background()
+			rows, err := ie.QueryIterator(itCtx, "sql-cursor", p.txn, statement)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to DECLARE CURSOR")
+			}
+			inputState := p.txn.GetLeafTxnInputState(ctx)
+			cursor := &sqlCursor{
+				InternalRows: rows,
+				readSeqNum:   inputState.ReadSeqNum,
+				txn:          p.txn,
+				statement:    statement,
+				created:      timeutil.Now(),
+			}
+			if err := p.sqlCursors.addCursor(cursorName, cursor); err != nil {
+				// This case shouldn't happen because cursor names are scoped to a session,
+				// and sessions can't have more than one statement running at once. But
+				// let's be diligent and clean up if it somehow does happen anyway.
+				_ = cursor.Close()
+				return nil, err
+			}
+			return newZeroNode(nil /* columns */), nil
+		},
+	}, nil
 }
 
 var errBackwardScan = pgerror.Newf(pgcode.ObjectNotInPrerequisiteState, "cursor can only scan forward")
@@ -219,7 +225,12 @@ func (f fetchNode) Close(ctx context.Context) {
 // CloseCursor implements the FETCH statement.
 // See https://www.postgresql.org/docs/current/sql-close.html for details.
 func (p *planner) CloseCursor(ctx context.Context, n *tree.CloseCursor) (planNode, error) {
-	return newZeroNode(nil /* columns */), p.sqlCursors.closeCursor(n.Name.String())
+	return &delayedNode{
+		name: n.String(),
+		constructor: func(ctx context.Context, p *planner) (planNode, error) {
+			return newZeroNode(nil /* columns */), p.sqlCursors.closeCursor(n.Name.String())
+		},
+	}, nil
 }
 
 type sqlCursor struct {

--- a/pkg/sql/sql_cursor_test.go
+++ b/pkg/sql/sql_cursor_test.go
@@ -1,0 +1,123 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sql
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/stretchr/testify/require"
+)
+
+// Make sure that preparing a DECLARE doesn't cause problems.
+func TestPrepareDeclare(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	srv, db, _ := serverutils.StartServer(t, base.TestServerArgs{Insecure: true})
+	defer srv.Stopper().Stop(context.Background())
+	defer db.Close()
+
+	ctx := context.Background()
+	conn, err := db.Conn(ctx)
+	require.NoError(t, err)
+
+	t.Run("prepare_declare_raw_txn", func(t *testing.T) {
+		// Make sure that preparing a DECLARE defers errors until execution.
+		stmt, err := conn.PrepareContext(ctx, "DECLARE foo CURSOR FOR VALUES (1), (2)")
+		require.NoError(t, err)
+
+		_, err = stmt.Exec()
+		require.EqualError(t, err, "pq: DECLARE CURSOR can only be used in transaction blocks")
+
+		// Make sure that we can use our prepared statement from before to
+		// successfully execute a declare cursor within a transaction.
+		// We need to execute a raw BEGIN so that we can reuse our pre-prepared txn.
+		_, err = conn.ExecContext(ctx, "BEGIN TRANSACTION")
+		require.NoError(t, err)
+
+		_, err = stmt.Exec()
+		require.NoError(t, err)
+
+		stmt, err = conn.PrepareContext(ctx, "FETCH 2 foo")
+		require.NoError(t, err)
+		r, err := stmt.Query()
+		require.NoError(t, err)
+		var actual int
+		r.Next()
+		require.NoError(t, r.Scan(&actual))
+		require.Equal(t, 1, actual)
+		more := r.Next()
+		require.Equal(t, true, more)
+		require.NoError(t, r.Scan(&actual))
+		require.Equal(t, 2, actual)
+		more = r.Next()
+		require.Equal(t, false, more)
+		_, err = conn.ExecContext(ctx, "COMMIT")
+		require.NoError(t, err)
+	})
+
+	t.Run("prepare_declare_driver_txn", func(t *testing.T) {
+		// Make sure that we can use the driver-level txn support to do the same thing.
+		tx, err := conn.BeginTx(context.Background(), nil /* opts */)
+		require.NoError(t, err)
+		stmt, err := tx.Prepare("DECLARE foo CURSOR FOR VALUES (1), (2)")
+		require.NoError(t, err)
+		_, err = stmt.Exec()
+		require.NoError(t, err)
+
+		stmt, err = tx.Prepare("FETCH 2 foo")
+		require.NoError(t, err)
+		r, err := stmt.Query()
+		require.NoError(t, err)
+		var actual int
+		r.Next()
+		require.NoError(t, r.Scan(&actual))
+		require.Equal(t, 1, actual)
+		more := r.Next()
+		require.Equal(t, true, more)
+		require.NoError(t, r.Scan(&actual))
+		require.Equal(t, 2, actual)
+		more = r.Next()
+		require.Equal(t, false, more)
+
+		require.NoError(t, tx.Commit())
+	})
+
+	// Make sure that we can use the automatic prepare support (when sending
+	// placeholders) to do the same thing.
+	// TODO (jordan): This currently doesn't work, because we don't fully walk
+	// the tree typechecking expressions when sending a DECLARE through the
+	// optimizer. See issue #77067.
+	// When this limitation is lifted, the rest of this test should be uncommented.
+	t.Run("prepare_declare_placeholder", func(t *testing.T) {
+		_, err = conn.ExecContext(ctx, "DECLARE foo CURSOR FOR SELECT 1 WHERE $1", true)
+		require.Contains(t, err.Error(), "could not determine data type of placeholder")
+
+		skip.WithIssue(t, 77067, "placeholders in DECLARE not supported")
+
+		stmt, err := conn.PrepareContext(ctx, "FETCH 1 foo")
+		require.NoError(t, err)
+		r, err := stmt.Query()
+		require.NoError(t, err)
+		var actual int
+		r.Next()
+		require.NoError(t, r.Scan(&actual))
+		require.Equal(t, 1, actual)
+		more := r.Next()
+		require.Equal(t, false, more)
+	})
+}

--- a/pkg/sql/sql_prepare_test.go
+++ b/pkg/sql/sql_prepare_test.go
@@ -18,12 +18,12 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
-	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 // Make sure that running a wire-protocol-level PREPARE of a SQL-level PREPARE
-// doesn't cause any problems.
-func TestPreparePrepare(t *testing.T) {
+// and SQL-level EXECUTE doesn't cause any problems.
+func TestPreparePrepareExecute(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
@@ -31,24 +31,32 @@ func TestPreparePrepare(t *testing.T) {
 	defer srv.Stopper().Stop(context.Background())
 	defer db.Close()
 
-	foo, err := db.Prepare("PREPARE x AS SELECT $1::int")
-	if err != nil {
-		t.Fatal(err)
-	}
-	_, err = foo.Exec()
-	if err != nil {
-		t.Fatal(err)
-	}
+	// Test that preparing an invalid EXECUTE fails at prepare-time.
+	_, err := db.Prepare("EXECUTE x(3)")
+	require.Contains(t, err.Error(), "no such prepared statement")
 
-	foo, err = db.Prepare("EXECUTE x(4)")
-	if err != nil {
-		t.Fatal(err)
-	}
-	r := foo.QueryRow()
+	// Test that we can prepare and execute a PREPARE.
+	s, err := db.Prepare("PREPARE x AS SELECT $1::int")
+	require.NoError(t, err)
+
+	_, err = s.Exec()
+	require.NoError(t, err)
+
+	// Make sure we can't send arguments to the PREPARE even though it has a
+	// placeholder inside (that placeholder is for the "inner" PREPARE).
+	_, err = s.Exec(3)
+	require.Contains(t, err.Error(), "expected 0 arguments, got 1")
+
+	// Test that we can prepare and execute the corresponding EXECUTE.
+	s, err = db.Prepare("EXECUTE x(3)")
+	require.NoError(t, err)
+
 	var output int
-	if err := r.Scan(&output); err != nil {
-		t.Fatal(err)
-	}
+	err = s.QueryRow().Scan(&output)
+	require.NoError(t, err)
+	require.Equal(t, 3, output)
 
-	assert.Equal(t, 4, output)
+	// Make sure we can't send arguments to the prepared EXECUTE.
+	_, err = s.Exec(3)
+	require.Contains(t, err.Error(), "expected 0 arguments, got 1")
 }

--- a/pkg/sql/sql_prepare_test.go
+++ b/pkg/sql/sql_prepare_test.go
@@ -1,0 +1,54 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sql
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/stretchr/testify/assert"
+)
+
+// Make sure that running a wire-protocol-level PREPARE of a SQL-level PREPARE
+// doesn't cause any problems.
+func TestPreparePrepare(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	srv, db, _ := serverutils.StartServer(t, base.TestServerArgs{Insecure: true})
+	defer srv.Stopper().Stop(context.Background())
+	defer db.Close()
+
+	foo, err := db.Prepare("PREPARE x AS SELECT $1::int")
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = foo.Exec()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	foo, err = db.Prepare("EXECUTE x(4)")
+	if err != nil {
+		t.Fatal(err)
+	}
+	r := foo.QueryRow()
+	var output int
+	if err := r.Scan(&output); err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Equal(t, 4, output)
+}

